node/network/src/proxy/mod.rs

```diff
@@ -0,0 +1,54 @@
+use std::sync::Arc;
+
+use super::message::Message;
+
+pub mod dropout;
+
+pub struct Proxy {
+    handlers: Vec<Arc<ProxyHandler>>
+}
+
+/// All messages before being sent to other participants will be processed by several handlers
+/// chained sequentially in a pipeline.
+///
+/// Each handler will receive an iterator of messages and will return (not necessarily the same)
+/// iterator of messages. This method allows easy manipulation of messages before being emitted
+/// to other authorities through the network.
+///
+/// The order in which handlers are grouped can affect the behavior of the proxy.
+impl Proxy {
+    pub fn new() -> Self {
+        Self {
+            handlers: vec![]
+        }
+    }
+
+    pub fn pipe<'a>(&'a self, message: &'a Message) -> Vec<&'a Message> {
+        let start_messages = vec![message];
+
+        self.handlers
+            .iter()
+            .fold(start_messages, |cur_messages, handler| {
+                handler.pipe(cur_messages)
+            })
+    }
+}
+
+/// ProxyHandler interface.
+pub trait ProxyHandler where Self: Send + Sync {
```

Comment:
I would also expect to have a FnProxyHandler - e.g. I pass function / closure and uses it for proxy. Very convenient for testing.

Reply:
Discussed in the #development channel.

Reply:
Moved to #799